### PR TITLE
Update Aspect/Fingerprint signatures

### PR DESCRIPTION
**Pull request by kipz**
- Diff.from is now optional so that initial fingerprints can be handled
- Handlers/workflows now take arrays of Diffs of same type so that aggregates can be calculated/handled within Aspects
- FP.type is now mandatory - we always need this, and this removes the ambiguity

### CODE REVIEW

**Comment by cdupuis** [comment]
Pull request rebase is in progress because @slimslenderslacks pushed 1 commit to **master**:
- cc3a9eb _upgrade clj-editors_

**Comment by cdupuis** [comment]
Pull request rebase is in progress because @slimslenderslacks pushed 1 commit to **master**:
- bf51b8e _Autofix: Third party licenses_

**Comment by cdupuis** [comment]
Pull request rebase is in progress because @atomist-bot pushed 1 commit to **master**:
- 4e8ef7c _Version: increment after 4.0.4 release_

**Comment by cdupuis** [comment]
Pull request rebase is in progress because @atomist-bot pushed 1 commit to **master**:
- 2c0eae6 _Changelog: add release 4.0.4_

**Comment by cdupuis** [comment]
Pull request rebase is in progress because @slimslenderslacks pushed 1 commit to **master**:
- f16718c _applyTargetFingerprint needs to be in the index if..._

File: lib/fingerprints/npmDeps.ts [hunk 1]
```diff
@@ -204,20 +204,18 @@ export const NpmCoordinates: Aspect = {
...
-                        diff.channel);
-                } else {
-                    return new Promise<Vote>(
-                        (resolve, reject) => {
-                            resolve({ abstain: true });
-                        },
-                    );
-                }
+            (ctx, diffs) => {
+                return Promise.all(_.map(diffs, diff => {
```
Comment (cdupuis) [inline lib/fingerprints/npmDeps.ts]: We never run things like this in a Promise.all as it provides no upper limit.

File: lib/machine/fingerprintSupport.ts [hunk 2]
```diff
@@ -101,16 +100,20 @@ export interface FingerprintImpactHandlerConfig {
...
-            aspect,
-            async () => {
-                return diff.targets;
-            },
-        );
-        return v;
+    async (ctx, diffs, aspect) => {
+        const abs = _.filter(diffs, diff => !diff.from);
+        return [
+            ... await Promise.all(
```
Comment (cdupuis) [inline lib/machine/fingerprintSupport.ts]: No Promise.all on unbounded collections.

File: lib/machine/fingerprintSupport.ts [hunk 3]
```diff
@@ -101,16 +100,20 @@ export interface FingerprintImpactHandlerConfig {
...
+                    _.map (
+                        _.difference(diffs, abs), diff => checkFingerprintTarget(
+                            ctx.context,
+                            diff,
+                            aspect,
+                            async () => diff.targets ))),
+            ... _.map(abs, () => {
+                return {abstain: true};
```
Comment (cdupuis) [inline lib/machine/fingerprintSupport.ts]: Format

File: lib/machine/fingerprintSupport.ts [hunk 4]
```diff
@@ -119,14 +122,14 @@ export const DefaultTargetDiffHandler: FingerprintDiffHandler =
...
-            return {
-                abstain: true,
-            };
-        }
+    return async (context, diffs, aspect) => {
+        const toDiff = _.filter(diffs, diff => diff.from && diff.to.sha !== diff.from.sha);
+        return [
+            ... await handler(context, toDiff, aspect),
+            ... _.map(_.difference(diffs, toDiff), () => {
+                return {abstain: true};
```
Comment (cdupuis) [inline lib/machine/fingerprintSupport.ts]: Format

File: lib/machine/fingerprintSupport.ts [hunk 5]
```diff
@@ -119,14 +122,14 @@ export const DefaultTargetDiffHandler: FingerprintDiffHandler =
...
-    return async (context, diff, aspect) => {
-        if (diff.from && diff.to.sha !== diff.from.sha) {
-            return handler(context, diff, aspect);
-        } else {
-            return {
-                abstain: true,
-            };
-        }
+    return async (context, diffs, aspect) => {
+        const toDiff = _.filter(diffs, diff => diff.from && diff.to.sha !== diff.from.sha);
```
Comment (cdupuis) [inline lib/machine/fingerprintSupport.ts]: Why lodash for a collection.filter?

File: lib/machine/runner.ts [hunk 6]
```diff
@@ -69,50 +69,55 @@ interface MissingInfo {
...
+        }
+        if (h.handler) {
+            voats.push(...await h.handler(i, diffs, aspect));
+        }
+        return voats;
+    })));
+
     if (aspect.workflows) {
-        aspectVotes = await Promise.all(aspect.workflows.map(h => h(i, diff, aspect)));
+        handlerVotes.push(... _.flatten(await Promise.all(_.map(aspect.workflows, wf => {
```
Comment (cdupuis) [inline lib/machine/runner.ts]: Remove Promise.all

File: lib/machine/runner.ts [hunk 7]
```diff
@@ -240,23 +245,23 @@ export function fingerprintRunner(
         }
         logger.info(`Found ${Object.keys(previous).length} fingerprints`);
 
-        const allFps: FP[] = await computer(fingerprinters, p);
+        const allFps = (await computer(fingerprinters, p));
```
Comment (cdupuis) [inline lib/machine/runner.ts]: Superfluous ()

File: lib/machine/runner.ts [hunk 8]
```diff
@@ -240,23 +245,23 @@ export function fingerprintRunner(
...
 
         try {
             const info = await missingInfo(i);
-            const allVotes: Vote[] = (await Promise.all(
-                allFps.map(fp => {
-                    const fpAspect: Aspect = fingerprinters.find(aspects => aspects.name === (fp.type || fp.name));
-                    return handleDiffs(fp, previous[fp.name], info, handlers, fpAspect, i);
+            const byType = _.groupBy(allFps, fp => fp.type);
+
+            const allVotes = _.flatten(await Promise.all(
```
Comment (cdupuis) [inline lib/machine/runner.ts]: Remove Promise.all

**Comment by cdupuis** [comment]
Running potentially side-effecting things in Promise.all has proven to be dangerous as node runs all passed callbacks in parallel.

File: lib/machine/fingerprintSupport.ts (hunk 3 above)
Comment (kipz) [reply lib/machine/fingerprintSupport.ts]: @cdupuis Thanks, but this seems like a job for a computer to me. Do we not have an auto-formatter available?

File: lib/machine/fingerprintSupport.ts (hunk 5 above)
Comment (kipz) [reply lib/machine/fingerprintSupport.ts]: Why not? I'm thinking consistency. Happy to use built in where available if that's preferred?

**Comment by kipz** [comment]
> Running potentially side-effecting things in Promise.all has proven to be dangerous as node runs all passed callbacks in parallel.

What's the alternative approach then please?

**Comment by kipz** [comment]
> > Running potentially side-effecting things in Promise.all has proven to be dangerous as node runs all passed callbacks in parallel.
> 
> What's the alternative approach then please?

oh - doing it one by one - i.e. chaining them?

File: lib/machine/fingerprintSupport.ts (hunk 5 above)
Comment (kipz) [reply lib/machine/fingerprintSupport.ts]: Also, this functional approach gives null safety, which is nice @cdupuis .

**Comment by cdupuis** [comment]
Pull request rebase is in progress because @johnsonr pushed 1 commit to **master**:
- eb3486a _Add type guard_

**Comment by cdupuis** [comment]
Pull request rebase is in progress because @slimslenderslacks pushed 2 commits to **master**:
- a63c732 _set success field in TransformResult to false when..._
- f9721c3 _Merge branch 'master' of github.com:atomist/sdm-pa..._

**Comment by cdupuis** [comment]
Pull request rebase is in progress because @johnsonr pushed 1 commit to **master**:
- 28c3c6b _Remove BaseAspect (<https://github.com/atomist/sdm-pack-fingerprints/issues/124|#124>)_

**Comment by cdupuis** [comment]
Pull request rebase is in progress because @johnsonr pushed 1 commit to **master**:
- 639a384 _Simplify typing of Aspect and remove unused member..._

**Pull request merged**